live-build/misc/upgrade-scripts/upgrade-container

```diff
@@ -273,12 +273,149 @@ function run() {
+		sed 's/[0-9]*$//'
+}
+
+function migrate_password_for_user() {
+	local user="$1"
+	local password
+
+	password="$(awk -F: "\$1 == \"$user\" {print \$2}" /etc/shadow)"
+	chroot "/var/lib/machines/$CONTAINER" usermod -p "$password" "$user" ||
+		die "'usermod -p ... $user' failed for '$CONTAINER'"
+}
+
+function migrate_file() {
+	local path="$1"
+	local directory
+
+	if [[ -f "$path" ]]; then
+		directory="$(dirname "$path")"
+		mkdir -p "/var/lib/machines/${CONTAINER}${directory}" ||
+			die "'mkdir -p $directory' failed for '$CONTAINER'"
+		cp "$path" "/var/lib/machines/${CONTAINER}${path}" ||
+			die "'cp $path' failed for '$CONTAINER'"
+	fi
+}
+
+function convert_to_bootfs_cleanup() {
+	for dir in /proc /sys /dev; do
+		umount -R "/var/lib/machines/${CONTAINER}${dir}" ||
+			warn "'umount -R' of '$dir' failed"
+	done
+}
+
+#
+# The purpose of this function is to convert an existing upgrade
+# container (specified by the $CONTAINER global variable) into the next
+# boot filesystem used by the appliance; i.e. after calling this
+# function, the container's root filesystem will be used as the
+# appliance's root filesystem, the next time the appliance boots.
+#
+# This is done by updating the appliance's bootloader (i.e. grub) to
+# point to the container's filesystem, along with setting the mountpoint
+# of the filesystem to be "/" instead of "/var/lib/machines/$CONTAINER".
+# This effectively removes the container, since systemd-nspawn only
+# looks in /var/lib/machines, so it'll no longer find the container
+# after the mountpoint changes.
+#
+function convert_to_bootfs() {
+	trap convert_to_bootfs_cleanup EXIT
+
+	for dir in /proc /sys /dev; do
+		mount --rbind "$dir" "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --rbind' of '$dir' failed"
+		mount --make-rslave "/var/lib/machines/${CONTAINER}${dir}" ||
+			die "'mount --make-rslave' of '$dir' failed"
+	done
+
+	#
+	# When performing a not-in-place upgrade, the root and delphix
+	# users will not have any password by default. Thus, we need to
+	# explicitly configure the passwords on the new root filesystem.
+	# Here, we ensure the passwords for the root and delphix user on
+	# the new root filesystem match their current values.
+	#
+	migrate_password_for_user delphix
+	migrate_password_for_user root
+
+	#
+	# LX-72 Until we have a proper solution for migrating the
+	# configuration of a given appliance to the new root filesystem
+	# (i.e. any configuration that isn't supplied by a package), we
+	# resort to explicitly migrating some of this configuration
+	# here. The files listed here are dynamically generated and/or
+	# modified by appliance-build.
+	#
+	while read -r file; do
+		migrate_file "$file"
+	done <<-EOF
+		/etc/cloud/cloud.cfg.d/99-delphix-development.cfg
+		/etc/cloud/cloud.cfg.d/99-delphix-internal.cfg
+		/etc/issue
+		/etc/nftables.conf
+		/etc/ssh/sshd_config
+		/etc/systemd/system/delphix-masking.service.d/override.conf
```

Comment:
How do we deal with services which are enabled by the stack and not enabled by default?

For example, snmpd and fluentd

Reply:
What do you mean, how is that handled on initial install? Presumably by the delphix-platform service? I’d have to look, but it should work the same as initial install.

Reply:
If we're doing a not-in-palce upgrade and fluentd was enabled by the stack, it looks like we would miss that configuration file on the newly upgraded dataset. Am I missing something?

Reply:
For example, I believe that fluentd creates an override file but I don't see it in the list

Reply:
Maybe the other way to look at it is, if delphix-platform handles this why do we need to copy over the override for these services explicitly here?

Reply:
Ah, yes. Anything done by the stack is going to be lost. That’ll need to be preserved by whatever mechanism we use for the “configuration migration” part. I’ve only added this migration as a short term solution, so the system will boot and be roughly configured the same as initial install.

Reply:
Just to be clear, the migration done here is purposefully incomplete. I only tried to add migration for things appliance-build currently does for the initial build/install so the system can boot properly after the upgrade. We’ll need to circle back and implement a more complete solution later.

Reply:
That addresses my concern. I was also wondering if we should use a migration.d directory to allow scripts to be placed in there to handle the migration of individual components. I know we've used this convention before so thought it might be useful here. It might be overkill too.

Reply:
Yea, this is basically what I was thinking as well. I haven’t thought of anything better. One drawback being, we can easily forget to add a migration when new code is added.

I plan to circle back to this later, and will seek design ideas from folks.